src/tokenizer.ts

```diff
@@ -432,26 +441,47 @@ export class Tokenizer {
                 break;
             // String
             case '"':
-                while (this.peek() != '"' && this.peek() != '\n' && !this.isAtEnd()) {
-                    this.advance();
-                }
-                if (this.peek() === '\n' || this.isAtEnd()) {
-                    throw new TokenizerErrors.UnterminatedStringError(this.line, this.col, this.source, this.start, this.current);
-                }
-                // Consume closing "
-                this.advance();
-                this.addStringToken(TokenType.STRING);
-                break;
-            case '\'':
-                while (this.peek() != '\'' && this.peek() != '\n' && !this.isAtEnd()) {
+            case "'":
+                let quote = c;
+                if (this.peek() == quote) { // second quote found
+                    this.advance(); // second quote consumed
+                    if (this.peek() != quote) { // empty string ""
+                        this.addStringToken(TokenType.STRING);
+                    } else if (this.peek() == quote) { // triple quotes found
+                        this.advance(); // third quote consumed
+                        while (this.peek() != quote && !this.isAtEnd()) {
+                            this.advance(); // advance until ending quote found
+                        }
+                        if (this.isAtEnd()) {
+                            throw new TokenizerErrors.UnterminatedStringError(this.line,
+                                this.col, this.source, this.start, this.current);
+                        }
+                        this.advance(); // consume first ending quote
+                        if (this.peek() == quote) {
+                            this.advance(); // consume second ending quote
+                            if (this.peek() == quote) {
+                                this.advance(); // consume third ending quote
+                                this.addMultiLineStringToken(TokenType.STRING);
+                            } else {
+                                throw new TokenizerErrors.UnterminatedStringError(this.line,
+                                    this.col, this.source, this.start, this.current);
+                            }
+                        } else {
+                            throw new TokenizerErrors.UnterminatedStringError(this.line,
+                                this.col, this.source, this.start, this.current);
+                        }
+                    }
+                } else {
+                    while (this.peek() != quote && this.peek() != '\n' && !this.isAtEnd()) {
+                        this.advance();
+                    }
+                    if (this.peek() === '\n' || this.isAtEnd()) {
+                        throw new TokenizerErrors.UnterminatedStringError(this.line, this.col, this.source, this.start, this.current);
+                    }
```

Comment:
So for this section, could you please use a different style of checks?

The idea is that we don't want to form highly-nested code (you'll learn more about this in CS2103T), so instead of
```
if (cond) {
} else {
}
```
Do
```
if (cond) {
# return early/ fail early
}
# Rest of the code
```

E.g.

```
                        if (this.peek() == quote) {
                            this.advance(); // consume second ending quote
                            if (this.peek() == quote) {
                                this.advance(); // consume third ending quote
                                this.addMultiLineStringToken(TokenType.STRING);
                            } else {
                                throw new TokenizerErrors.UnterminatedStringError(this.line,
                                    this.col, this.source, this.start, this.current);
                            }
                        } else {
                            throw new TokenizerErrors.UnterminatedStringError(this.line,
                                this.col, this.source, this.start, this.current);
                        }
```
should be (notice how much more linear and less nested the code is!)
```
                        if (this.peek() != quote) {
                            throw new TokenizerErrors.UnterminatedStringError(this.line,
                                this.col, this.source, this.start, this.current);
                        }
                        this.advance(); // consume second ending quote
                        if (this.peek() != quote) {
                            throw new TokenizerErrors.UnterminatedStringError(this.line,
                                this.col, this.source, this.start, this.current);
                        }
		        this.advance(); // consume third ending quote
		        this.addMultiLineStringToken(TokenType.STRING);

```
please apply this to the whole section.